src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs

```diff
@@ -112,363 +113,991 @@ public class ControllerActionInvoker : IActionInvoker
+                        var task = InvokeNextExceptionFilterAsync();
+                        if (task.Status != TaskStatus.RanToCompletion)
+                        {
+                            next = State.ExceptionSyncEnd;
+                            return task;
+                        }
+
+                        goto case State.ExceptionSyncEnd;
+                    }
+
+                case State.ExceptionSyncEnd:
+                    {
+                        var filter = (IExceptionFilter)state;
+                        var exceptionContext = _exceptionContext;
+
+                        if (exceptionContext?.Exception != null)
+                        {
+                            _diagnosticSource.BeforeOnException(exceptionContext, filter);
+
+                            // Exception filters only run when there's an exception - unsetting it will short-circuit
+                            // other exception filters.
+                            filter.OnException(exceptionContext);
+
+                            _diagnosticSource.AfterOnException(exceptionContext, filter);
+
+                            if (exceptionContext.Exception == null)
+                            {
+                                _logger.ExceptionFilterShortCircuited(filter);
+                            }
+                        }
+
+                        goto case State.ExceptionEnd;
+                    }
+
+                case State.ExceptionInside:
+                    {
+                        goto case State.ActionBegin;
+                    }
+
+                case State.ExceptionShortCircuit:
+                    {
+                        Debug.Assert(_exceptionContext != null);
+
+                        Task task;
+                        if (scope == Scope.Resource)
+                        {
+                            Debug.Assert(_exceptionContext.Result != null);
+                            _resourceExecutedContext = new ResourceExecutedContext(_controllerContext, _filters)
+                            {
+                                Result = _exceptionContext.Result,
+                            };
+
+                            task = InvokeResultAsync(_exceptionContext.Result);
+                            if (task.Status != TaskStatus.RanToCompletion)
+                            {
+                                next = State.ResourceEnd;
+                                return task;
+                            }
+
+                            goto case State.ResourceEnd;
+                        }
+
+                        task = InvokeResultAsync(_exceptionContext.Result);
+                        if (task.Status != TaskStatus.RanToCompletion)
+                        {
+                            next = State.InvokeEnd;
+                            return task;
+                        }
+
+                        goto case State.ResourceEnd;
+                    }
+
+                case State.ExceptionEnd:
```

Comment:
Need to think what the right behavior is when an exception was handled by short-circuiting (setting Exception = null) but no result was set. (Need to take into account old MVC behavior)